### PR TITLE
Use GetLastErrorMsg more

### PR DESCRIPTION
**Pull request by JosJuice**
It's thread safe (unlike strerror), and it lets us get error strings (rather than just numbers) on Windows without needing to write extra lines of code.

### CODE REVIEW

**Comment by JosJuice** [comment]
Oh, great... We have a pre-existing problem in GetLastErrorMsg that wasn't caught until now. I'm not sure how it should be fixed.

**Comment by BhaaLseN** [comment]
Is it a good idea to do successive calls to `GetLastError` and `GetLastErrorMsg` like that? There could be plenty of things that change or reset the error being returned.

Other than that, the Linux fails probably need a CMake flag or define to fix this: http://stackoverflow.com/a/3052083

> From the header `string.h`:
> 
```
/* Reentrant version of `strerror'.
   There are 2 flavors of `strerror_r', GNU which returns the string
   and may or may not use the supplied temporary buffer and POSIX one
   which fills the string into the buffer.
   To use the POSIX version, -D_XOPEN_SOURCE=600 or -D_POSIX_C_SOURCE=200112L
   without -D_GNU_SOURCE is needed, otherwise the GNU version is
   preferred.  */
```
> 
> Note, be careful when using GNU extensions, turn them on (_GNU_SOURCE) last, before including the headers that you want it to affect (or undefine it strategically). No need to worry if not using GNU extensions, though.
> 
> Generally, if GNU deviates from POSIX in default behavior, you'll see some comments in the header to indicate how you can get the POSIX behavior. Its also (usually) documented in the glibc manual, but that doesn't always make it to the highly condensed man pages.

**Comment by sepalani** [comment]
I'm pretty sure that change is really UB on Windows. As ```GetLastError()``` should be called right after the error occurred. For instance, creating a string will reset errno to 0 and will output the message string "Success". 

The only alternative I can see is a ```GetLastErrorAndMessage()``` that returns a tuple with the error code and its message that called/saved errorno/GetLastError() in first place as we have no guarantee with the current C++ version used concerning the parameters evaluation order.

Otherwise, an optional parameter for ```GetLastErrorMessage()``` (that would more likely become ```GetErrorMessage()```) is a solution if ```GetLastError()``` was called before.

File: Source/Core/Common/FileUtil.cpp [hunk 1]
```diff
@@ -149,13 +149,13 @@ bool CreateDir(const std::string& path)
...
-  DWORD error = GetLastError();
-  if (error == ERROR_ALREADY_EXISTS)
+  if (GetLastError() == ERROR_ALREADY_EXISTS)
   {
     WARN_LOG(COMMON, "CreateDir: CreateDirectory failed on %s: already exists", path.c_str());
     return true;
   }
-  ERROR_LOG(COMMON, "CreateDir: CreateDirectory failed on %s: %i", path.c_str(), error);
+  ERROR_LOG(COMMON, "CreateDir: CreateDirectory failed on %s: %s", path.c_str(),
+            GetLastErrorMsg().c_str());
```
Comment (shuffle2) [inline Source/Core/Common/FileUtil.cpp]: Generally you should not call GetLastError more than once for a given cause (even if you know _currently_ the code cannot do something to set the last error to something else in between).

File: Source/Core/Core/HW/WiimoteReal/IOWin.cpp [hunk 2]
```diff
@@ -591,21 +591,18 @@ int IORead(HANDLE& dev_handle, OVERLAPPED& hid_overlap_read, u8* buf, int index)
...
         // In case it was aborted by someone else
-        if (ERROR_OPERATION_ABORTED == overlapped_err)
+        if (ERROR_OPERATION_ABORTED == GetLastError())
         {
           return -1;
         }
 
-        WARN_LOG(WIIMOTE, "GetOverlappedResult error %d on Wiimote %i.", overlapped_err, index + 1);
+        WARN_LOG(WIIMOTE, "GetOverlappedResult error on Wiimote %i: %s", index + 1,
+                 GetLastErrorMsg().c_str());
```
Comment (shuffle2) [inline Source/Core/Core/HW/WiimoteReal/IOWin.cpp]: same here

**Comment by shuffle2** [comment]
Oh I see now @sepalani basically said the same thing. overall the PR looks fine, but yea - the api should be changed to only fetch the error once (asap). The same would benefit non-windows.

**Comment by JosJuice** [comment]
The problems with calling GetLastError too late should be fixed now. I'm still not sure exactly what to do about the GNU problem, though.

**Comment by sepalani** [comment]
@JosJuice If you can't get around this error using CMake, is this trick working?
```C++
#ifdef _GNU_SOURCE
#undef _GNU_SOURCE
#include <cstring>
#define _GNU_SOURCE
#else
#include <cstring>
#endif

```

**Comment by JosJuice** [comment]
It does seem to work. Thanks! (I haven't tried doing anything about it in CMake, though.)

File: Source/Core/Common/CommonFuncs.cpp [hunk 3]
```diff
@@ -2,32 +2,59 @@
...
 #endif
 
-// Generic function to get last error message.
-// Call directly after the command or use the error num.
-// This function might change the error code.
-std::string GetLastErrorMsg()
+int GetLastErrorCode()
+{
+#ifdef _WIN32
+  return GetLastError();
```
Comment (shuffle2) [inline Source/Core/Common/CommonFuncs.cpp]: This is kind of misleading: Windows still has `errno`, and it's used for the same CRT/POSIX-y things as on non-windows. `GetLastError` is for win32 apis.

Comment (JosJuice) [inline Source/Core/Common/CommonFuncs.cpp]: So the same applies to GetLastErrorMsg, right? We want to get messages in the win32 way if the error code is from `GetLastError` and we want to get messages in the POSIX-y way if the error code is from `errno`? If so, I think I'll end up redoing this PR in a different way...

Comment (shuffle2) [inline Source/Core/Common/CommonFuncs.cpp]: Yes, it seems sometimes the functions are being called on windows in order to report an `errno`. (just skimmed, but for example in FileUtil.cpp for fseek).

Comment (shuffle2) [inline Source/Core/Common/CommonFuncs.cpp]: It's kind of a side effect of these Common files being a huge #ifdef stew.

**Pull request closed without merging**